homework_1/02_ProgressiveTax/taxes/taxes.go

```diff
@@ -0,0 +1,75 @@
+package taxes
+
+import (
+	"sort"
+
+	"github.com/pkg/errors"
+)
+
+type bracket struct {
+	threshold float32
+	taxRate   float32
+}
+
+func NewBracket(threshold, taxRate float32) (*bracket, error) {
+	if threshold < 0 {
+		return nil, errors.New("bracket threshold is negative")
+	}
+
+	if taxRate < 0 {
+		return nil, errors.New("bracket tax rate is negative")
+	}
+
+	return &bracket{threshold: threshold, taxRate: taxRate}, nil
+}
+
+func CreateBrackets(thresholds, taxRates []float32) ([]*bracket, error) {
+	if len(thresholds) != len(taxRates) {
+		return nil, errors.New("thresholds and taxRates do not have the same length")
+	}
+
+	var brackets []*bracket
+
+	for idx, threshold := range thresholds {
+		bracket, err := NewBracket(threshold, taxRates[idx])
+
+		if err != nil {
+			return nil, err
+		}
+
+		brackets = append(brackets, bracket)
+	}
+
+	return brackets, nil
+}
```

Comment:
Did you think about a solution where you don't have `CreateBrackets` functions but you just expose `bracket` as public and let the user directly modify that struct? You can validate the input brackets at the start of `CalculateProgressiveTax` then. I'm not saying your implementation is wrong, you've made `bracket` private which, essentially, forces the user to use `CreateBrackets` function just curious about the reasoning.

Reply:
I provided `NewBracket` and `CreateBrackets` as public to serve as some sort of constructors. If we would, for example, have to create brackets on multiple places, this makes the process easier and in one place.